src/meta-client/src/client/ddl.rs

```diff
@@ -0,0 +1,199 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use api::v1::meta::ddl_task_client::DdlTaskClient;
+use api::v1::meta::heartbeat_client::HeartbeatClient;
+use api::v1::meta::{
+    AskLeaderRequest, ErrorCode, RequestHeader, Role, SubmitDdlTaskRequest, SubmitDdlTaskResponse,
+};
+use common_grpc::channel_manager::ChannelManager;
+use common_telemetry::debug;
+use snafu::{ensure, OptionExt, ResultExt};
+use tokio::sync::RwLock;
+use tonic::transport::Channel;
+
+use crate::client::Id;
+use crate::error;
+use crate::error::Result;
+
+#[derive(Clone, Debug)]
+// TODO(weny): removes this in following PRs.
+#[allow(unused)]
+pub struct Client {
+    inner: Arc<RwLock<Inner>>,
+}
+
+// TODO(weny): removes this in following PRs.
+#[allow(dead_code)]
+impl Client {
+    pub fn new(id: Id, role: Role, channel_manager: ChannelManager) -> Self {
+        let inner = Arc::new(RwLock::new(Inner {
+            id,
+            role,
+            channel_manager,
+            peers: vec![],
+            leader: None,
+        }));
+
+        Self { inner }
+    }
+
+    pub async fn start<U, A>(&mut self, urls: A) -> Result<()>
+    where
+        U: AsRef<str>,
+        A: AsRef<[U]>,
+    {
+        let mut inner = self.inner.write().await;
+        inner.start(urls).await
+    }
+
+    pub async fn is_started(&self) -> bool {
+        let inner = self.inner.read().await;
+        inner.is_started()
+    }
+
+    pub async fn submit_ddl_task(
+        &self,
+        req: SubmitDdlTaskRequest,
+    ) -> Result<SubmitDdlTaskResponse> {
+        let mut inner = self.inner.write().await;
+        inner.submit_ddl_task(req).await
+    }
+}
+
+#[derive(Debug)]
+// TODO(weny): removes this in following PRs.
+#[allow(unused)]
+struct Inner {
+    id: Id,
+    role: Role,
+    channel_manager: ChannelManager,
+    peers: Vec<String>,
+    leader: Option<String>,
+}
+
+impl Inner {
+    async fn start<U, A>(&mut self, urls: A) -> Result<()>
+    where
+        U: AsRef<str>,
+        A: AsRef<[U]>,
+    {
+        ensure!(
+            !self.is_started(),
+            error::IllegalGrpcClientStateSnafu {
+                err_msg: "Router client already started",
+            }
+        );
+
+        self.peers = urls
+            .as_ref()
+            .iter()
+            .map(|url| url.as_ref().to_string())
+            .collect::<HashSet<_>>()
+            .drain()
+            .collect::<Vec<_>>();
+
+        Ok(())
+    }
+
+    // TODO(weny): considers refactoring `ask_leader` into a common client.
+    async fn ask_leader(&mut self) -> Result<()> {
```

Comment:
can we make heartbeat client a field of this ddl client? so ddl client could just use the ask_leader method in heartbeat client

Reply:
Maybe we can add an API for HeartbeatClient: `heartbeat_client.get_leader() -> Peer` based on `ask_leader()`, then:
```
let leader = heartbeat_client.get_leader();
ddl.client.submit_task(leader);
```